### PR TITLE
add basic support of route handlers

### PR DESCRIPTION
**Pull request by DmytroHryshyn**

### CODE REVIEW

File: next/13/replace-api-routes/index.ts [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+		.getStatements()
+		.find((s) => Node.isExportAssignment(s));
+
+	const declarations =
+		exportAssignment
+			?.getFirstDescendantByKind(SyntaxKind.Identifier)
+			?.getSymbol()
+			?.getDeclarations() ?? [];
+
+	let handler:
```
Comment (grzpab) [inline next/13/replace-api-routes/index.ts:60]: `let handler : Handler | null = null`

File: next/13/replace-api-routes/index.ts [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+		| ArrowFunction
+		| FunctionExpression
+		| FunctionDeclaration
+		| undefined;
+
+	declarations.forEach((d) => {
```
Comment (grzpab) [inline next/13/replace-api-routes/index.ts:66]: we always take the last handler here, right?

File: next/13/replace-api-routes/index.ts [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+const HTTP_METHODS = ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'] as const;
+
+export type HTTPMethod = (typeof HTTP_METHODS)[number];
+
+const RESPONSE_INIT_FIELDS = ['headers', 'status', 'statusText'] as const;
+
+type ResponseInitParam = (typeof RESPONSE_INIT_FIELDS)[number];
+type ResponseInit = Partial<{ [k in ResponseInitParam]: unknown }>;
+
+const unquotify = (input: string): string => {
```
Comment (grzpab) [inline next/13/replace-api-routes/index.ts:103]: what was the reason for that function's existence?

File: next/13/replace-api-routes/index.ts [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+	return expression.getName();
+};
+
+const rewriteResponseCallExpressions = (handler: Handler) => {
+	const responseInit: ResponseInit = {};
+
+	const callExpressions = handler
+		.getDescendantsOfKind(SyntaxKind.CallExpression)
+		.filter((callExpression) => {
```
Comment (grzpab) [inline next/13/replace-api-routes/index.ts:130]: it can be inline function

File: next/13/replace-api-routes/index.ts [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+			const name = getCallExpressionName(childCallExpression);
+
+			if (RESPONSE_INIT_FIELDS.includes(name as ResponseInitParam)) {
+				responseInit[name as ResponseInitParam] = unquotify(
+					childCallExpression.getArguments()[0]?.getText() ?? '',
+				);
+			}
+		});
+
+		const callExpressionArg =
```
Comment (grzpab) [inline next/13/replace-api-routes/index.ts:149]: you can move this above line 139 and use in line 144

File: next/13/replace-api-routes/index.ts [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+		handler.insertStatements(
+			idx,
+			`return NextResponse.json(${callExpressionArg}, ${JSON.stringify(
+				responseInit,
+			)})`,
+		);
+	});
+};
+
+// const rewriteReqResImports = (sourceFile: SourceFile) => {
```
Comment (grzpab) [inline next/13/replace-api-routes/index.ts:163]: is this still needed?

File: next/13/replace-api-routes/index.ts [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+					);
+				}
+			}
+		}
+	});
+
+	const positionAfterImports = getPositionAfterImports(sourceFile);
+
+	Array.from(HTTPMethodHandlers).forEach(([method, methodHandler]) => {
+		const [statement] = sourceFile.insertStatements(
```
Comment (grzpab) [inline next/13/replace-api-routes/index.ts:249]: should we check if statement exists?

File: next/13/replace-api-routes/index.ts (hunk 3 above)
Comment (DmytroHryshyn) [reply next/13/replace-api-routes/index.ts:103]: sometimes `getText` returns a string with extra quotes like `'next/server'` or `"next/server"`, so i have to remove them

File: next/13/replace-api-routes/index.ts (hunk 6 above)
Comment (DmytroHryshyn) [reply next/13/replace-api-routes/index.ts:163]: removed

File: next/13/replace-api-routes/index.ts (hunk 5 above)
Comment (DmytroHryshyn) [reply next/13/replace-api-routes/index.ts:150]: line 144 is inside loop of childCallExpressions  and `callExpressionArg` is a parent call expression arg. so i think it should be here

File: next/13/replace-api-routes/index.ts (hunk 3 above)
Comment (DmytroHryshyn) [reply next/13/replace-api-routes/index.ts]: fixed

File: next/13/replace-api-routes/index.ts (hunk 1 above)
Comment (DmytroHryshyn) [reply next/13/replace-api-routes/index.ts]: fixed

**Review by grzpab — APPROVED** [review]

**Pull request merged**